### ISSUE TITLE
error: [SERVER] Port 3000 is already in use! starting Node V2.x after Migrating from Node V1.x  - Is it possible my instance of Node v2.x is using wrong background agent?

### ISSUE THREAD

**Issue by ArlenNY** [Requarks/wiki#2110]
**Host Info (please complete the following information):**
OS: Red Hat Enterprise Linux Server 7.7 (Maipo)
Wiki.js version: 2.4.107 (Migrated from from 1.0.78)
Database engine: PostgreSQL 12.3 (working on migrating from MongoDB 3.4.18)

**Question**
Thanks for your assistance with bug #2096. I installed the latest stable version of PostgreSQL (12.10.x)  and created a new database on the same server as the old database to migrate our old db data from. The path of the wiki v2.x installation is /wiki/

When I try to launch node using node server, I get the output below showing port 3000 is already in use so I investigated using netstat and found node running on this port already. I verified that the old version of node (v1.x) is no longer installed, killed the pid (1564) before retrying to start node again and the process appeared again on another pid (2814) I seem to have issue. I did some google searching and came across  https://github.com/Requarks/wiki-v1/issues/100 on Wiki V1. Are these possibly related? I could be wrong but node could be calling /root/server/agent.js from the Wiki.js v1 installation and would need it to call the Wiki V2 background agent: 

[root@wiki server]# cat /root/server/agent.js
// ===========================================
// Wiki.js - Background Agent
// 1.0.0
// Licensed under AGPLv3
// ===========================================
 

I'm looking into the issue and will continue tomorrow morning. Thank you in advance. 

---

[root@wiki ~]# node server
2020-06-26T02:48:10.220Z - info: [SERVER] Wiki.js is initializing...
2020-06-26T02:48:10.222Z - info: [SERVER] /root
2020-06-26T02:48:10.333Z - info: [SERVER] Checking data directories...
2020-06-26T02:48:10.335Z - info: [SERVER] Data and Repository directories are OK.
2020-06-26T02:48:10.546Z - info: [SERVER] Checking Git repository...
2020-06-26T02:48:11.623Z - info: [SERVER] Starting HTTP/WS server on port 3000...
2020-06-26T02:48:11.645Z - error: [SERVER] Port 3000 is already in use!
[root@wiki ~]# 2020-06-26T02:48:11.757Z - info: [AGENT] Background Agent is initializing...
2020-06-26T02:48:12.285Z - info: [AGENT] Checking Git repository...
2020-06-26T02:48:12.856Z - warn: [AGENT] Lost connection to main server. Exiting...
/root/server/agent.js:222
  job.stop()
      ^

TypeError: Cannot read property 'stop' of undefined
    at process.<anonymous> (/root/server/agent.js:222:7)
    at process.emit (events.js:315:20)
    at finish (internal/child_process.js:861:14)
    at processTicksAndRejections (internal/process/task_queues.js:79:11)

[root@wiki ~]# netstat -tunapl
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:5432          0.0.0.0:*               LISTEN      1694/postmaster
tcp        0      0 127.0.0.1:25            0.0.0.0:*               LISTEN      1929/master 
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      1957/nginx: master
tcp        0      0 127.0.0.1:65443         0.0.0.0:*               LISTEN      1140/nailsd 
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      1957/nginx: master
tcp        0      0 0.0.0.0:55443           0.0.0.0:*               LISTEN      1174/nailswebd
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1662/sshd   
tcp        0      0 127.0.0.1:60206         127.0.0.1:65443         ESTABLISHED 1159/mon    
tcp        0     64 10.104.204.116:22       192.168.87.83:61211     ESTABLISHED 2192/sshd: root@pts
tcp        0      0 127.0.0.1:65443         127.0.0.1:60206         ESTABLISHED 1144/nailslogd
tcp6       0      0 :::3000                 :::*                    LISTEN      1564/node /root/ser
tcp6       0      0 ::1:5432                :::*                    LISTEN      1694/postmaster
tcp6       0      0 ::1:25                  :::*                    LISTEN      1929/master 
tcp6       0      0 :::443                  :::*                    LISTEN      1957/nginx: master
tcp6       0      0 :::3306                 :::*                    LISTEN      1807/mysqld 
tcp6       0      0 :::80                   :::*                    LISTEN      1957/nginx: master
tcp6       0      0 :::22                   :::*                    LISTEN      1662/sshd   
udp        0      0 127.0.0.1:323           0.0.0.0:*                           1073/chronyd
udp6       0      0 ::1:48372               ::1:48372               ESTABLISHED 1694/postmaster
udp6       0      0 :::8082                 :::*                                1176/macmnsvc
udp6       0      0 ::1:323                 :::*                                1073/chronyd
[root@wiki ~]# kill 1564
[root@wiki ~]# netstat -tunapl
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:5432          0.0.0.0:*               LISTEN      1694/postmaster
tcp        0      0 127.0.0.1:25            0.0.0.0:*               LISTEN      1929/master 
tcp        0      0 0.0.0.0:443             0.0.0.0:*               LISTEN      1957/nginx: master
tcp        0      0 127.0.0.1:65443         0.0.0.0:*               LISTEN      1140/nailsd 
tcp        0      0 0.0.0.0:80              0.0.0.0:*               LISTEN      1957/nginx: master
tcp        0      0 0.0.0.0:55443           0.0.0.0:*               LISTEN      1174/nailswebd
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1662/sshd   
tcp        0      0 127.0.0.1:60206         127.0.0.1:65443         ESTABLISHED 1159/mon    
tcp        0     64 10.104.204.116:22       192.168.87.83:61211     ESTABLISHED 2192/sshd: root@pts
tcp        0      0 127.0.0.1:65443         127.0.0.1:60206         ESTABLISHED 1144/nailslogd
tcp6       0      0 :::3000                 :::*                    LISTEN      2814/node /root/ser
tcp6       0      0 ::1:5432                :::*                    LISTEN      1694/postmaster
tcp6       0      0 ::1:25                  :::*                    LISTEN      1929/master 
tcp6       0      0 :::443                  :::*                    LISTEN      1957/nginx: master
tcp6       0      0 :::3306                 :::*                    LISTEN      1807/mysqld 
tcp6       0      0 :::80                   :::*                    LISTEN      1957/nginx: master
tcp6       0      0 :::22                   :::*                    LISTEN      1662/sshd   
udp        0      0 127.0.0.1:323           0.0.0.0:*                           1073/chronyd
udp6       0      0 ::1:48372               ::1:48372               ESTABLISHED 1694/postmaster
udp6       0      0 :::8082                 :::*                                1176/macmnsvc
udp6       0      0 ::1:323                 :::*                                1073/chronyd

[root@wiki ~]# rpm -q --whatprovides /root/.pm2/pm2.log
file /root/.pm2/pm2.log is not owned by any package
[root@wiki ~]# vi /root/.pm2/pm2.log
2020-06-25 21:18:56: Time                 : Thu Jun 25 2020 21:18:56 GMT-0400 (Eastern Daylight Time)
2020-06-25 21:18:56: PM2 version          : 2.6.1
2020-06-25 21:18:56: Node.js version      : 12.18.1
2020-06-25 21:18:56: Current arch         : x64
2020-06-25 21:18:56: PM2 home             : /root/.pm2
2020-06-25 21:18:56: PM2 PID file         : /root/.pm2/pm2.pid
2020-06-25 21:18:56: RPC socket file      : /root/.pm2/rpc.sock
2020-06-25 21:18:56: BUS socket file      : /root/.pm2/pub.sock
2020-06-25 21:18:57: Application log path : /root/.pm2/logs
2020-06-25 21:18:57: Process dump file    : /root/.pm2/dump.pm2
2020-06-25 21:18:57: Concurrent actions   : 2
2020-06-25 21:18:57: SIGTERM timeout      : 1600
2020-06-25 21:18:57: ===============================================================================
2020-06-25 21:18:57: Starting execution sequence in -fork mode- for app name:wiki id:0
2020-06-25 21:18:57: App name:wiki id:0 online
2020-06-25 22:47:53: App [wiki] with id [0] and pid [1564], exited with code [0] via signal [SIGINT]
2020-06-25 22:47:53: Starting execution sequence in -fork mode- for app name:wiki id:0
2020-06-25 22:47:53: App name:wiki id:0 online

**Comment by Smankusors** [comment]
If you open port 3000 on the browser, which shows? Wiki.js v1.x or v2.x? Or neither?

**Comment by ArlenNY** [comment]
Neither will show because node won't start. Only wiki.js 2.x rpms are on the server now.

**Comment by ArlenNY** [comment]
Any of the *awesome* developers on the Wiki.js Dev team have any ideas on this?

**Comment by Smankusors** [comment]
huh? wait wait, where do you install Wiki.js 2.x? 😨

Maybe you should run `node server` when current directory is `/root`? 🤔

**Comment by ArlenNY** [comment]
Thanks Smakusors, the Wiki.js V2 app installation directory is  /wiki so when I stop the Wiki.js V1 application using "pm2 stop wiki", change directory to /wiki and run the command "node server"....the Wiki.js V2 attempts to start but hangs at the step below. I waited for ten minutes before using Ctrl+c to stop. 

Can anyone provide assistance or hints?

`[root@wiki wiki]# node server
Loading configuration from /wiki/config.yml... OK
2020-09-24T16:32:39.864Z [MASTER] info: =======================================
2020-09-24T16:32:39.865Z [MASTER] info: = Wiki.js 2.4.107 =====================
2020-09-24T16:32:39.865Z [MASTER] info: =======================================
2020-09-24T16:32:39.865Z [MASTER] info: Initializing...
2020-09-24T16:32:47.925Z [MASTER] info: Using database driver pg for postgres [ OK ]
2020-09-24T16:32:47.928Z [MASTER] info: Connecting to database...
2020-09-24T16:32:48.316Z [MASTER] info: Database Connection Successful [ OK ]
2020-09-24T16:33:02.721Z [MASTER] warn: Mail is not setup! Please set the configuration in the administration area!
2020-09-24T16:33:04.386Z [MASTER] info: Loading GraphQL Schema...
2020-09-24T16:33:09.511Z [MASTER] info: GraphQL Schema: [ OK ]
2020-09-24T16:33:09.704Z [MASTER] info: HTTP Server on port: [ 80 ]
2020-09-24T16:33:10.004Z [MASTER] warn: listen EADDRINUSE: address already in use 0.0.0.0:80
2020-09-24T16:33:10.257Z [MASTER] info: No new analytics providers found: [ SKIPPED ]
2020-09-24T16:33:10.564Z [MASTER] info: No new authentication strategies found: [ SKIPPED ]
2020-09-24T16:33:10.614Z [MASTER] info: No new comment providers found: [ SKIPPED ]
2020-09-24T16:33:10.678Z [MASTER] info: No new editors found: [ SKIPPED ]
2020-09-24T16:33:10.826Z [MASTER] info: No new loggers found: [ SKIPPED ]
2020-09-24T16:33:11.112Z [MASTER] info: No new renderers found: [ SKIPPED ]
2020-09-24T16:33:11.229Z [MASTER] info: No new search engines found: [ SKIPPED ]
2020-09-24T16:33:11.360Z [MASTER] info: No new storage targets found: [ SKIPPED ]
2020-09-24T16:33:11.361Z [MASTER] info: Checking for installed optional extensions...
2020-09-24T16:33:11.426Z [MASTER] info: Optional extension git is installed. [ OK ]
2020-09-24T16:33:11.472Z [MASTER] info: Optional extension pandoc was not found on this system. [ SKIPPED ]
2020-09-24T16:33:11.485Z [MASTER] info: Optional extension puppeteer was not found on this system. [ SKIPPED ]
2020-09-24T16:33:11.492Z [MASTER] info: Optional extension sharp was not found on this system. [ SKIPPED ]
2020-09-24T16:33:11.541Z [MASTER] info: Authentication Strategy local: [ OK ]
2020-09-24T16:33:19.058Z [MASTER] info: (COMMENTS/DEFAULT) Initializing...
2020-09-24T16:33:19.058Z [MASTER] info: (COMMENTS/DEFAULT) Initialization completed.
2020-09-24T16:33:19.469Z [MASTER] info: (STORAGE/DISK) Initializing...
2020-09-24T16:33:19.470Z [MASTER] info: (STORAGE/DISK) Initialization completed.
2020-09-24T16:33:19.480Z [MASTER] info: Purging orphaned upload files...
2020-09-24T16:33:19.488Z [MASTER] info: Syncing locales with Graph endpoint...
2020-09-24T16:33:19.496Z [MASTER] info: Fetching latest updates from Graph endpoint...
2020-09-24T16:33:19.519Z [MASTER] info: Purging orphaned upload files: [ COMPLETED ]
Loading configuration from /wiki/config.yml... OK
2020-09-24T16:33:19.777Z [MASTER] info: Fetching latest updates from Graph endpoint: [ COMPLETED ]
2020-09-24T16:33:20.089Z [MASTER] info: Pulled latest locale updates for English from Graph endpoint: [ COMPLETED ]
2020-09-24T16:33:20.093Z [MASTER] info: Syncing locales with Graph endpoint: [ COMPLETED ]
2020-09-24T16:33:20.239Z [JOB] info: Rebuilding page tree...
2020-09-24T16:33:20.626Z [JOB] info: Using database driver pg for postgres [ OK ]
2020-09-24T16:33:20.943Z [JOB] info: Rebuilding page tree: [ COMPLETED ]`

**Comment by Smankusors** [comment]
well it hangs because it's not doing anything. It's idle. I mean, it's ready to open in your browser.

**Comment by ArlenNY** [comment]
I used netstat to confirm nginx was running on port 80 as shown above. I stopped nginx, ran `node server` command again, and the Wiki.js v2 application started. 

Now that I've gotten Wiki.js V2 application to start, does anyone know how to reconfigure pm2 to start the Wiki.js V2 application on boot? My pm2 is currently configured to start the old Wiki.js V1 application and I need to migrate away from the Wiki.js V1 and run Wiki.js V2 full time. Thanks in advance for your help!
